controllers/csm_controller.go

```diff
@@ -556,28 +556,45 @@ func (r *ContainerStorageModuleReconciler) oldStandAloneModuleCleanup(ctx contex
 			}
 		}
 		// check if observability needs to be uninstalled
-		// NOTE: check if individual component need to be uninstalled
-		oldObservabilityEnabled, _ := utils.IsModuleEnabled(ctx, *oldCR, csmv1.Observability)
+		oldObservabilityEnabled, oldObs := utils.IsModuleEnabled(ctx, *oldCR, csmv1.Observability)
 		newObservabilityEnabled, _ := utils.IsModuleEnabled(ctx, *newCR, csmv1.Observability)
-		if oldObservabilityEnabled && !newObservabilityEnabled {
+		// check if observability components need to be uninstalled
+		components := []string{}
+		if oldObservabilityEnabled && newObservabilityEnabled {
+			for _, comp := range oldObs.Components {
+				oldCompEnabled := utils.IsComponentEnabled(ctx, *oldCR, csmv1.Observability, comp.Name)
+				newCompEnabled := utils.IsComponentEnabled(ctx, *newCR, csmv1.Observability, comp.Name)
+				if oldCompEnabled && !newCompEnabled {
+					components = append(components, comp.Name)
+				}
+			}
+		}
+		if (oldObservabilityEnabled && !newObservabilityEnabled) || len(components) > 0 {
 			_, clusterClients, err := utils.GetDefaultClusters(ctx, *oldCR, r)
 			if err != nil {
 				return err
 			}
 			for _, cluster := range clusterClients {
 				// remove module observability
 				log.Infow("Deleting observability")
-				if err = r.reconcileObservability(ctx, true, operatorConfig, *oldCR, cluster.ClusterCTRLClient, cluster.ClusterK8sClient); err != nil {
+				if err = r.reconcileObservability(ctx, true, operatorConfig, *oldCR, components, cluster.ClusterCTRLClient, cluster.ClusterK8sClient); err != nil {
 					return err
 				}
 			}
-
 		}
 	}
 
-	annotations := newCR.GetAnnotations()
-	newCR.Annotations[previouslyAppliedCustomResource] = newCR.Annotations["kubectl.kubernetes.io/last-applied-configuration"]
-	newCR.SetAnnotations(annotations)
+	copyCR := newCR.DeepCopy()
+	delete(copyCR.Annotations, previouslyAppliedCustomResource)
+	delete(copyCR.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
+	copyCR.ManagedFields = nil
+	copyCR.Status = csmv1.ContainerStorageModuleStatus{}
+	out, err := json.Marshal(copyCR)
+	if err != nil {
+		return fmt.Errorf("error marshalling CR to annotation: %v", err)
+	}
+	newCR.Annotations[previouslyAppliedCustomResource] = string(out)
```

Comment:
@taohe1012 
1. why copyCR.ManagedFields need to be nil?
2. why copyCR.Status is an empty struct?
3. what's the difference between previously-applied-configuration and last-applied-configuration?

Reply:
@YianZong 
1. because it is useless, and if not set to nil, ManagedFields is long json and the annotation looks too long;
2. copyCR.Statusis useless, and in the step of oldStandAloneModuleCleanup, the state is not final state, so make it empty;
3.  previously-applied-configuration is our self defined annotation.
"kubectl.kubernetes.io/last-applied-configuration" is control by 'kubectl apply -f [files]' action, which has some limitations(as below). 
  1) Its content will be updated only when the csm is modified by 'kubectl apply -f [files]'
  2) If creating CR by 'kubectl create -f', the last-applied-configuration content will be empty.
  3) and if modify csm by 'kubectl edit csm ', the last-applied-configuration content won't be updated; 
So now use previously-applied-configuration directly, and deprecated the last-applied-configuration.